### PR TITLE
Prepare input objects

### PR DESCRIPTION
**Pull request by theorygeek**
This allows you to build input objects for, eg, range types, and get a `Range` passed into your app.

Example:
```ruby
class DateRangeInput < GraphQL::Schema::InputObject
  argument :min, Int, required: true
  argument :max, Int, required: true

  def prepare
    return unless min <= max
    min..max
  end
end
```

Resolves https://github.com/rmosolgo/graphql-ruby/issues/903#issuecomment-424113851

### CODE REVIEW

File: lib/graphql/query/literal_input.rb [hunk 1]
```diff
@@ -108,7 +108,13 @@ def self.from_arguments(ast_arguments, argument_owner, variables)
           end
         end
 
-        argument_owner.arguments_class.new(values_hash, context: context, defaults_used: defaults_used)
+        result = argument_owner.arguments_class.new(values_hash, context: context, defaults_used: defaults_used)
+
+        if result.respond_to?(:prepare)
```
Comment (eapache) [inline lib/graphql/query/literal_input.rb]: should the base classes for all of these just `def prepare; self; end`?

**Comment by theorygeek** [comment]
A couple of challenges with this implementation:

1. I tried to also give you a nice validation error if your `prepare` returns nil, similar to the way that `coerce_non_null_input` on scalar types behaves. However, it doesn't work for literal input; I'm not entirely sure what's missing there, or the best way to rectify it. 

(Probably too much duplicated code for handling input objects?)

2. This introduces a performance regression, since as-implemented, it will instantiate each input object _twice_:
- Once during `validate_non_null_input` inside of GraphQL::InputObjectType, to check for `nil`'s
- And then again when the value is actually being passed into a resolver (the exact mechanics here is a bit of a mystery to me, though)

**Comment by eapache** [comment]
I'm a bit surprised this doesn't break something when it tries to recurse into `DateRangeInput` in order to prepare/validate `min` and `max` but can't because it not longer an `InputObject`?

**Comment by theorygeek** [comment]
> I'm a bit surprised this doesn't break something when it tries to recurse
I think that recursion happens inside of the `#initialize` method on GraphQL::Schema::InputObject.

File: spec/graphql/schema/input_object_spec.rb [hunk 2]
```diff
@@ -89,12 +89,81 @@ class Schema < GraphQL::Schema
...
+      query_str = <<-GRAPHQL
+      query ($input: InputObj!){ inputs(input: $input) }
+      GRAPHQL
+
+      res = InputObjectPrepareTest::Schema.execute(query_str, variables: { input: { min: 5, max: 10 } })
+      expected_obj = (5..10).inspect
+      assert_equal expected_obj, res["data"]["inputs"]
+    end
+    
+    it "gives an error if prepare returns nil (literal)" do
```
Comment (theorygeek) [inline spec/graphql/schema/input_object_spec.rb]: This test isn't passing, but I'm not entirely sure what's missing.

File: lib/graphql/input_object_type.rb [hunk 3]
```diff
@@ -147,6 +153,11 @@ def validate_non_null_input(input, ctx)
         end
       end
 
+      # Better way to do this, perhaps keeping the coerced value around?
+      if coerce_non_null_input(input, ctx).nil?
```
Comment (theorygeek) [inline lib/graphql/input_object_type.rb]: Introduces a performance regression: input is converted into the `InputObject` class created by the user, and the custom `prepare` method is invoked, but all of that work is thrown away.

If the user's `initialize` and `prepare` methods are expensive, it'll cause a problem. Also, they might not expect those methods to get invoked twice: once for validation, and then once for the actual value.

**Comment by theorygeek** [comment]
I eliminated the use-case for dynamic, runtime validation (eg, by returning `nil` from a `prepare` method). The tests are passing, except for a seemingly unrelated one (something for `GraphQL::InternalRepresentation::Rewrite`).

I also eliminated the `respond_to?` checks, and updated the guides. Would love some feedback from @rmosolgo on this.

**Comment by rmosolgo** [comment]
Sorry to leave this hanging for a year 😞 . Are you still interested in a change like this? 

I was initially turned off by it, because argument `prepare:` hooks have been nothing but a headache for me. But honestly, this approach seems really straightforward, and if you have a use for it, I'm game to merge it. 

I've tried to resolve the conflicts 🤞

**Comment by theorygeek** [comment]
> Are you still interested in a change like this?

Yeah, I still think this would still be valuable.

**Pull request merged**